### PR TITLE
Add support for secrets in package configurations

### PR DESCRIPTION
**Pull request by kritika-singh3**
Description: Added support to specify secrets in package configurations meaning both in package repository and the package.
 - the rules will be evaluated and resolved during
    - check connection call
    - while saving package configurations
    - while checking for material update (if the package is used as a material in any pipeline)
 - users can specify the rules for a secret config using `package_repository` as entity
 - the secrets in packages will be validated based upon the rules for the package repository.

### CODE REVIEW

File: config/config-api/src/main/java/com/thoughtworks/go/domain/packagerepository/PackageDefinition.java [hunk 1]
```diff
@@ -326,4 +322,22 @@ public void ensureIdExists() {
             setId(UUID.randomUUID().toString());
         }
     }
+
+    @Override
+    public boolean hasSecretParams() {
+        return this.getRepository().hasSecretParams()
+                || this.getConfiguration()
+                .stream()
+                .anyMatch(ConfigurationProperty::hasSecretParams);
```
Comment (maheshp) [inline config/config-api/src/main/java/com/thoughtworks/go/domain/packagerepository/PackageDefinition.java:331]: The logic for `hasSecretParams` and `getSecretParams` is repeated across classes. Is it ok to move it to `Configuration` class?

File: server/src/main/java/com/thoughtworks/go/server/service/RulesService.java [hunk 2]
```diff
@@ -136,24 +143,22 @@ private void addError(Map<CaseInsensitiveString, StringBuilder> pipelinesWithErr
...
 
-    @NotNull
-    private HashMap<CaseInsensitiveString, StringBuilder> validate(SCM scmConfig) {
-        String scmConfigName = scmConfig.getName();
+    private String errorString(Map<CaseInsensitiveString, StringBuilder> errors) {
+        return join(errors.values(), '\n').trim();
+    }
+
+    protected Map<CaseInsensitiveString, StringBuilder> validate(SecretParams secretParams, Class<? extends Validatable> entityClass, String entityName, String entityNameOrErrorMessagePrefix) {
         HashMap<CaseInsensitiveString, StringBuilder> pipelinesWithErrors = new HashMap<>();
```
Comment (maheshp) [inline server/src/main/java/com/thoughtworks/go/server/service/RulesService.java:168]: `Map<CaseInsensitiveString, StringBuilder> errors = new HashMap<>();`

**Review by maheshp — COMMENTED** [review]
Mostly looks ok to me, I will test it before merging once  the changes are made.

**Comment by maheshp** [comment]
@kritika-singh3 came across couple of issues,
1. Secrets for package materials are not resolved before agent assignment, hence it fails.
2. While configuring packages, any errors during secret resolution returns a 500 and appropriate errors are not shown to the user.

**Comment by kritika-singh3** [comment]
@maheshp Pushed the fix for the scenarios you mentioned. 👍

File: server/src/main/java/com/thoughtworks/go/server/service/materials/PackageDefinitionService.java [hunk 3]
```diff
@@ -138,6 +145,9 @@ private void update(Username username, PackageDefinition packageDeinition, HttpL
         } catch (Exception e) {
             if (e instanceof GoConfigInvalidException && !result.hasMessage()) {
                 result.unprocessableEntity(entityConfigValidationFailed(packageDeinition.getClass().getAnnotation(ConfigTag.class).value(), packageDeinition.getId(), e.getMessage()));
+            } else if (e instanceof RulesViolationException || e instanceof SecretResolutionFailureException) {
+                LOGGER.error(e.getMessage(), e);
```
Comment (maheshp) [inline server/src/main/java/com/thoughtworks/go/server/service/materials/PackageDefinitionService.java:149]: Don't log this error.

File: server/src/main/java/com/thoughtworks/go/server/service/materials/PackageRepositoryService.java [hunk 4]
```diff
@@ -173,6 +176,9 @@ private void update(Username username, HttpLocalizedOperationResult result, Enti
         } catch (Exception e) {
             if (e instanceof GoConfigInvalidException && !result.hasMessage()) {
                 result.unprocessableEntity(entityConfigValidationFailed(repository.getClass().getAnnotation(ConfigTag.class).value(), repository.getId(), e.getMessage()));
+            } else if (e instanceof RulesViolationException || e instanceof SecretResolutionFailureException) {
+                LOGGER.error(e.getMessage(), e);
```
Comment (maheshp) [inline server/src/main/java/com/thoughtworks/go/server/service/materials/PackageRepositoryService.java:180]: Don't log this error.

**Review by maheshp — APPROVED** [review]

**Comment by maheshp** [comment]
Verified on `GoCD Version: 20.8.0 (12211-5d6b4a7a7145f728857afb8108b31effbbd1cd0c)`

**Pull request merged**